pkg/controller/attach_detach.go

```diff
@@ -0,0 +1,586 @@
+		if err != nil {
+			klog.Warningf("failed to complete initial AzVolumeAttachment sync: %v", err)
+		}
+		if !retry {
+			break
+		}
+	}
+
+	return nil
+}
+
+// ManageAttachmentsForVolume will be running on a separate channel
+func (r *ReconcileAttachDetach) syncAll(ctx context.Context, syncedVolumeAttachments map[string]bool, volumesToSync map[string]bool) (bool, map[string]bool, map[string]bool, error) {
+	// Get all volumeAttachments
+	volumeAttachments, err := r.kubeClient.StorageV1().VolumeAttachments().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return true, syncedVolumeAttachments, volumesToSync, err
+	}
+
+	if syncedVolumeAttachments == nil {
+		syncedVolumeAttachments = map[string]bool{}
+	}
+	if volumesToSync == nil {
+		volumesToSync = map[string]bool{}
+	}
+
+	// Loop through volumeAttachments and create Primary AzVolumeAttachments in correspondence
+	for _, volumeAttachment := range volumeAttachments.Items {
+		// skip if sync has been completed volumeAttachment
+		if syncedVolumeAttachments[volumeAttachment.Name] {
+			continue
+		}
+		if volumeAttachment.Spec.Attacher == azureutils.DriverName {
+			volumeName := volumeAttachment.Spec.Source.PersistentVolumeName
+			if volumeName == nil {
+				continue
+			}
+			// get PV and retrieve diskName
+			pv, err := r.kubeClient.CoreV1().PersistentVolumes().Get(ctx, *volumeName, metav1.GetOptions{})
+			if err != nil {
+				klog.Errorf("failed to get PV (%s): %v", *volumeName, err)
+				return true, syncedVolumeAttachments, volumesToSync, err
+			}
+
+			if pv.Spec.CSI == nil || pv.Spec.CSI.Driver != azureutils.DriverName {
+				continue
+			}
+			volumesToSync[pv.Spec.CSI.VolumeHandle] = true
+
+			diskName, err := azureutils.GetDiskNameFromAzureManagedDiskURI(pv.Spec.CSI.VolumeHandle)
+			if err != nil {
+				klog.Warningf("failed to extract disk name from volumehandle (%s): %v", pv.Spec.CSI.VolumeHandle, err)
+				delete(volumesToSync, pv.Spec.CSI.VolumeHandle)
+				continue
+			}
+			nodeName := volumeAttachment.Spec.NodeName
+			azVolumeAttachmentName := azureutils.GetAzVolumeAttachmentName(diskName, nodeName)
+
+			// check if the CRI exists already
+			azVolumeAttachment, err := azureutils.GetAzVolumeAttachment(ctx, r.client, r.azVolumeClient, azVolumeAttachmentName, r.namespace, false)
+			klog.Infof("Recovering AzVolumeAttachment(%s)", azVolumeAttachmentName)
+			// if CRI already exists, append finalizer to it
+			if err == nil {
+				azVolumeAttachment = r.initializeMeta(ctx, azVolumeAttachment)
+				err = r.update(ctx, azVolumeAttachment)
+				if err != nil {
+					klog.Errorf("failed to add finalizer to AzVolumeAttachment (%s): %v", azVolumeAttachmentName, err)
+					return true, syncedVolumeAttachments, volumesToSync, err
+				}
+				// if not found, create one
+			} else if errors.IsNotFound(err) {
+				azVolumeAttachment := v1alpha1.AzVolumeAttachment{
+					ObjectMeta: metav1.ObjectMeta{
```

Comment:
Shouldn't we add the AzVolumeAttachment finalizers for this newly created CRI too?

Reply:
You are right. Thanks for catching that.